metricshub-wbem-extension/src/main/java/org/sentrysoftware/metricshub/extension/wbem/WbemExtension.java

```diff
@@ -0,0 +1,389 @@
+					metricFactory.collectNumberMetric(hostMonitor, WBEM_UP_METRIC, UP, strategyTime);
+					return;
+				}
+				log.debug(
+					"Hostname {} - Checking WBEM protocol status. WBEM exception when performing a WQL SELECT query on '{}' namespace: ",
+					hostname,
+					wbemNamespace,
+					e
+				);
+			}
+		}
+
+		// Generate a metric from the WBEM result
+		// CHECKSTYLE:OFF
+		metricFactory.collectNumberMetric(hostMonitor, WBEM_UP_METRIC, wbemResult != null ? UP : DOWN, strategyTime);
+		// CHECKSTYLE:ON
+	}
+
+	@Override
+	public CriterionTestResult processCriterion(
+		Criterion criterion,
+		String connectorId,
+		TelemetryManager telemetryManager
+	) {
+		if (criterion instanceof WbemCriterion wbemCriterion) {
+			// Sanity check
+			if (criterion == null) {
+				return CriterionTestResult.error(wbemCriterion, "Malformed criterion. Cannot perform detection.");
+			}
+
+			// Gather the necessary info on the test that needs to be performed
+			final String hostname = telemetryManager.getHostConfiguration().getHostname();
+
+			final WbemConfiguration wbemConfiguration = (WbemConfiguration) telemetryManager
+				.getHostConfiguration()
+				.getConfigurations()
+				.get(WbemConfiguration.class);
+			if (wbemConfiguration == null) {
+				return CriterionTestResult.error(wbemCriterion, "The WBEM credentials are not configured for this host.");
+			}
+
+			WbemCriterionProcessor wbemCriterionProcessor = new WbemCriterionProcessor(wbemRequestExecutor);
+
+			// If namespace is specified as "Automatic"
+			if (AUTOMATIC_NAMESPACE.equalsIgnoreCase(wbemCriterion.getNamespace())) {
+				final String cachedNamespace = telemetryManager
+					.getHostProperties()
+					.getConnectorNamespace(connectorId)
+					.getAutomaticWbemNamespace();
+
+				// If not detected already, find the namespace
+				if (cachedNamespace == null) {
+					return wbemCriterionProcessor.findNamespace(
+						hostname,
+						wbemConfiguration,
+						wbemCriterion,
+						telemetryManager,
+						connectorId
+					);
+				}
+
+				// Update the criterion with the cached namespace
+				WqlCriterion cachedNamespaceCriterion = wbemCriterion.copy();
+				cachedNamespaceCriterion.setNamespace(cachedNamespace);
+
+				// Run the test
+				return wbemCriterionProcessor.performDetectionTest(
+					hostname,
+					wbemConfiguration,
+					cachedNamespaceCriterion,
+					telemetryManager
+				);
+			}
+
+			// Run the test
+			return wbemCriterionProcessor.performDetectionTest(hostname, wbemConfiguration, wbemCriterion, telemetryManager);
+		}
+
+		throw new IllegalArgumentException(
+			String.format(
+				"Hostname %s - Cannot process criterion %s.",
+				telemetryManager.getHostname(),
+				criterion != null ? criterion.getClass().getSimpleName() : "<null>"
+			)
+		);
+	}
+
+	public SourceTable processSource(Source source, String connectorId, TelemetryManager telemetryManager) {

```

Comment:
Same as previous, I was expecting to see:
```java
		if (source instanceof WbemSource wbemSource) {
			return new WbemSourceProcessor(wbemRequestExecutor).process(wbemSource, connectorId, telemetryManager);
		}
		throw new IllegalArgumentException(
			String.format(
				"Hostname %s - Cannot process source %s.",
				telemetryManager.getHostname(),
				source != null ? source.getClass().getSimpleName() : "<null>"
			)
		);
```
Move all the logic in the WbemSourceProcessor.